crates/ty_python_semantic/src/types/infer.rs

```diff
@@ -4660,6 +4660,49 @@ impl<'db, 'ast> TypeInferenceBuilder<'db, 'ast> {
         }
     }
 
+    fn infer_global_statement(&mut self, global: &ast::StmtGlobal) {
+        // CPython allows examples like this, where a global variable is never explicitly defined
+        // in the global scope:
+        //
+        // ```py
+        // def f():
+        //     global x
+        //     x = 1
+        // def g():
+        //     print(x)
+        // ```
+        //
+        // However, we consider that an unresolved reference error, and we require the variable to
+        // be explicitly defined (either bound or declared) in the global scope.
+        let ast::StmtGlobal {
+            node_index: _,
+            range,
+            names,
+        } = global;
+        let global_place_table = self.index.place_table(FileScopeId::global());
+        for name in names {
+            if let Some(place_id) = global_place_table.place_id_by_name(name) {
+                let place = global_place_table.place_expr(place_id);
+                if place.is_bound() || place.is_declared() {
+                    // This name is explicitly defined in the global scope (not just in function
+                    // bodies that mark it `global`).
+                    continue;
+                }
+            }
+            if !builtins_symbol(self.db(), name).place.is_unbound() {
+                // This name is either the name of a a built-in (like `int`) or an implicit global
+                // symbol (like `__file__`).
+                continue;
+            }
```

Comment:
I'm not sure we should exclude builtins from this check: declaring a symbol as `global` in an inner scope and assigning to it from that inner scope will insert a symbol into the global scope that will shadow the symbol from the builtin scope, greatly complicating ty's ability to accurately infer boundness and types of symbols in the global scope. Implicit globals _should_ be excluded from this check, though, as they _are_ defined directly in the module's global scope.